### ISSUE TITLE
twisted.python.deprecate allows no indication of what something has been deprecated in favor of

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3047]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3047|
|Type|enhancement|
|Created|2008-02-21 17:16:49Z|
|Branch|https://github.com/twisted/twisted/tree/deprecated-replacement-3047|

Often a deprecated API is superseded by another.  Using the decorator available in `twisted.python.deprecate` doesn't allow what the new, preferred API to be specified in the deprecation message, though.

CompatibilityPolicy isn't complete or official yet (though I'm not sure why), but it points out that the first steps of introducing a deprecation are:

    * New functionality is introduced.
    * Old functionality is deprecated with a `PendingDeprecationWarning` which mentions release R+2 as the release it is slated for completion, and _explains the location of the new functionality_. 

(emphasis mine)

Some code is using this decorator already, but at the time of the filing of this ticket, it is actually used for functionality which has _no_ replacement, so this may not really be a release critical issue.  It may make sense to take it out of the Twisted 8.0 milestone before the release, rather than blocking on it.

<details><summary>Searchable metadata</summary>

```
trac-id__3047 3047
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_deprecated_replacement_3047 branches/deprecated-replacement-3047
branch_author__therve__glyph therve, glyph
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1203614209000000 1203614209000000
changetime__1297656310000000 1297656310000000
version__None None
owner__glyph glyph
cc__jesstess cc__mesozoic
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_1"></a>|@jml commented|
|-|-|

I don't think that this is release critical. 'replacement' can be added as an optional parameter at any time before or after the release.

I think that being able to specify a replacement is, in addition to complying to a policy on a wiki,  a useful feature.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_2"></a>|@jml commented|
|-|-|

Also, I think that for simplicity, we should forget that `PendingDeprecationWarning` exists.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve commented|
|-|-|

(In [23530]) Branching to 'calldeprecated-message-3047'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve commented|
|-|-|

(In [23531]) Add the message parameter, and tests.

Refs [#3047](https://github.com/twisted/twisted/issues/3047)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve removed owner|
|-|-|

OK, I think this is ready to review. It addresses the main plot of the ticket, which is the ability to mention what's the replacement. I think the ability to make `deprecated` uses `PendingDeprecationWarning` (if it should done) should be do separately: we first need a way to know the difference between 2 versions, then we could manipulate this to change `deprecated`.

Other note: it's unfortunate that t.p.deprecate is public. We should try to keep these modules private until there is a need of an external usage.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @therve|
|-|-|

Indeed, this goes towards what I wanted fixed when I filed this ticket.  Also, I agree that it would have been better if we had kept t.p.deprecate to ourselves, at least for a while.  Oh well.

As for what's in the branch, how would you feel about making this change slightly more specialized?  What if `"please use %s instead."` were inside `deprecate` (or one of the helpers) and only the value to be interpolated into it were specified as a parameter?  This would keep thinking for developers deprecating APIs to a minimum, provide more consistency in the messages, and maybe even give pydoctor a chance to link to the new API automatically.  `callDeprecated` could even try to verify that it actually exists. >:)  I haven't looked over all of our uses of `deprecate` so I'm not sure this covers all our cases, but it seems like it should.  If it doesn't, then I guess we shouldn't do it. ;)  OTOH, there could be a couple different ways to specify the deprecation message, with a common one only requiring the new API name.

It looks like the signature of `callDeprecated` could get pretty confusing if we decide to make the `deprecated` API much more complicated.  I wonder if there's some way to avoid that.  If `deprecated` were a class instead, with attributes like `version` and `message` (or whatever) it'd be easy to just pass one of those in to `callDeprecated`.  Maybe that's too big of a change, though?  I'm not sure, mostly just brainstorming now.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph set owner to @glyph<br>@glyph set status to `assigned`|
|-|-|

See my comment on [#4804](https://github.com/twisted/twisted/issues/4804)

**Comment by twisted-trac** [comment]
|[<img alt="mesozoic's avatar" src="https://avatars.githubusercontent.com/u/458347?s=50" width="50" height="50">](https://github.com/mesozoic)<a name="note_11"></a>|@mesozoic commented|
|-|-|

In [30504], I changed the ```message``` parameter to be ```replacement```, which can take a callable or a string. Not all tests are passing on Python 2.6 because this branch doesn't contain the fix for [#3223](https://github.com/twisted/twisted/issues/3223). Ready for review.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_12"></a>|@glyph commented|
|-|-|

(In [30615]) Branching to 'deprecated-replacement-3047'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_13"></a>|@glyph commented|
|-|-|

(In [30619]) don't refer to the test itself in the test docstring; just say what the behavior is re [#3047](https://github.com/twisted/twisted/issues/3047)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_14"></a>|@glyph commented|
|-|-|

(In [30624]) Monster docstring upgrade for callDeprecated, explaining all of its behavior and
preferring the new (version, replacement) argument format.  re [#3047](https://github.com/twisted/twisted/issues/3047)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_15"></a>|@glyph commented|
|-|-|

Unfortunately I just took over authorship and committed all of my feedback rather than reviewing.  So hopefully this should be easy for the next reviewer, but I don't think I can review it now having authored so much.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_16"></a>|@glyph removed owner<br>@glyph set status to `new`|
|-|-|

[Build Results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/deprecated-replacement-3047)
[Merged Diff Text](http://twistedmatrix.com:12435/3047)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun set owner to @glyph|
|-|-|

Thanks.  Here's some feedback, _exclusively_ about whitespace:

  1. There are some misformatted docstrings
    1. a couple parts of `_getDeprecationWarningString` got flowed together (`format` particularly)
    1. there's some extra vertical whitespace in `_getDeprecationDocstring`
    1. `getDeprecationWarningString` has an extra `@return` halfway through
    1. `callDeprecated` also has some extra vertical whitespace
  1. `test_getDeprecationWarningStringReplacementWithCallable` has some trailing whitespace
  1. The indentation being asserted by `test_deprecatedReplacement` and `test_deprecatedReplacementWithCallable` is a little gross, isn't it?  Can we do better or is it insanely hard?

If the last point isn't too ridiculous, feel free to merge once these are addressed.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_19"></a>|@glyph commented|
|-|-|

(In [30746]) review point 1.1 re [#3047](https://github.com/twisted/twisted/issues/3047)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_20"></a>|@glyph commented|
|-|-|

(In [30747]) review point 1.2 re [#3047](https://github.com/twisted/twisted/issues/3047)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_21"></a>|@glyph commented|
|-|-|

(In [30748]) review point 1.3 re [#3047](https://github.com/twisted/twisted/issues/3047)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_22"></a>|@glyph commented|
|-|-|

(In [30749]) review point 2 re [#3047](https://github.com/twisted/twisted/issues/3047)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_23"></a>|@glyph commented|
|-|-|

As to point 1.4: No it doesn't, I don't know what you mean.

As to point 3: the actual indentation is fine (note that there are no '\n's in front of the non-indented lines).  It's just that quoted, inside the test case, it's hard to wrap things in such a way that it looks reasonable.

I shall merge forthwith.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_24"></a>|@glyph set status to `closed`|
|-|-|

(In [30751]) Added a 'replacement' parameter to the 'twisted.python.deprecate.deprecated'
decorator.  This allows deprecations to unambiguously specify what they have
been deprecated in favor of.

Author: therve, mesozoic, glyph

Reviewer: glyph, exarkun

Fixes: [#3047](https://github.com/twisted/twisted/issues/3047)
